### ISSUE TITLE
Data transfer between GPU and Proxy

### ISSUE THREAD

**Issue by fyf2016** [NVIDIA/nccl#852]
Hi，
I'm reading the NCCL source code, but what confuses me is that I don't know when the GPU data is ready and passed to the Proxy. If two GPUs are transmitted in Net mode by sharing Host mem, what is the transmission process between GPU-Proxy and Proxy-GPU?

My understanding is that the GPU will first map a piece of memory to the CPU memory of Host mem, and then when the GPU prepares the data, it will first send the data to the shared memory of Host mem, and then transmit it to another GPU through the network.
I don't know how the Proxy knows that the GPU has data ready, and where in the source code it puts the data into the buffer and sends it to the GPU.

**Comment by fyf2016** [comment]
Suppose I want to communicate between GPU1 and GPU2. I apply for a buffer buff1 on GPU1, which contains the data I store. Next, according to the net communication method of nccl, I need to transfer the data in buff1 to the shared buffer area buff0 (there is a mapping relationship between buff and the cpu memory of Host mem). 
But how does nccl transfer the data from buff1 to buff0?

**Comment by sjeaugey** [comment]
That's done through the netSendProgress/netRecvProgress functions calling the network plugin isend/irecv functions?

**Comment by jiangxiaobin96** [comment]
When i use ncclSend function, i will define a block of dev memory which contain send data. If i want to send data to an other gpu, nccl needs to transfer the data to shared memory and send data from shared memory using netSendProgress. So where transfer the address user defined to shared memory address?

**Comment by fyf2016** [comment]
<img width="1090" alt="image" src="https://github.com/NVIDIA/nccl/assets/46273019/a823009c-3d54-4472-a952-061856094a80">
Why send data from buff to GPU first in the `sendProxyProgress` function? Where does the data in this buff come from?

**Comment by sjeaugey** [comment]
This steps is providing the buffer address to the GPU, so that the GPU knows where to write data to. It's only providing a pointer, not the data.

**Comment by fyf2016** [comment]
Is the buffer address the address that the GPU maps to the CPU buffer in Host mem? (If the two GPUs use Net mode, communicate through Host mem)

**Comment by sjeaugey** [comment]
Yes, that is correct. If we use GPU Direct RDMA, that address is in GPU memory, otherwise it is in CPU memory (allocated through cudaHostAlloc).

**Comment by fyf2016** [comment]
OK～ Thanks a lot for helping me to solve my doubts.

**Comment by fyf2016** [comment]
Wait a minute, I hava last question, how the data transfer to mapped buffer in GPU? Is it using the StoreLL and readLL commands? Are there other commands involved?

**Comment by fyf2016** [comment]
Wait a minute, I hava last question, how the data transfer to mapped buffer in GPU? Is it using the StoreLL and readLL commands? Are there other commands involved?

**Comment by ZhiyiHu1999** [comment]
> Yes, that is correct. If we use GPU Direct RDMA, that address is in GPU memory, otherwise it is in CPU memory (allocated through cudaHostAlloc).

Hello, @sjeaugey. Does sendproxyprogress and recvproxyprogress only include net mode (GPU-CPU(proxy)-net-CPU(proxy)-GPU) or both net mode and RDMA (GPU-net-GPU) are included in sendproxyprogress and recvproxyprogress? Thanks a lot!

**Comment by sjeaugey** [comment]
> Wait a minute, I hava last question, how the data transfer to mapped buffer in GPU? Is it using the StoreLL and readLL commands? Are there other commands involved?

StoreLL is only used for the LL protocol. For the simple protocol, the `reduceCopy` functions perform loads and stores.

> Does sendproxyprogress and recvproxyprogress only include net mode (GPU-CPU(proxy)-net-CPU(proxy)-GPU) or both net mode and RDMA (GPU-net-GPU) are included in sendproxyprogress and recvproxyprogress

As mentioned before, the only difference with GPU Direct RDMA is that the buffer is in GPU memory instead of CPU memory. All the code is the same.

**Comment by ZhiyiHu1999** [comment]
> > Wait a minute, I hava last question, how the data transfer to mapped buffer in GPU? Is it using the StoreLL and readLL commands? Are there other commands involved?
> 
> StoreLL is only used for the LL protocol. For the simple protocol, the `reduceCopy` functions perform loads and stores.
> 
> > Does sendproxyprogress and recvproxyprogress only include net mode (GPU-CPU(proxy)-net-CPU(proxy)-GPU) or both net mode and RDMA (GPU-net-GPU) are included in sendproxyprogress and recvproxyprogress
> 
> As mentioned before, the only difference with GPU Direct RDMA is that the buffer is in GPU memory instead of CPU memory. All the code is the same.

Hello, @sjeaugey . Thanks a lot! I have two more questions:
(1) To my limited understanding, the cuda kernel is responsible for moving the data from user defined sending buffer to shared buffer between GPUs, then sendProxyProgress is responsible for moving the data in the shared buffer from sender to receiver through the network. Finally, the cuda kernel moves the data from shared buffer to receiving buffer. Is the procedure correct and if so, what is recvProxyProgress calling irecv responsible for? In brief, I am not sure isend/irecv moves data from where to where.

(2) The sendProxyProgress calls isend when there are jobs in the circular FIFO shared buffer but how does recvProxyProgress determine to call irecv?

Thanks a lot for your kind help!

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're [reviewing](https://github.com/NVIDIA/nccl/discussions/1761) older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
